### DISCUSSION TITLE
Choose versioning scheme and next version number

### DISCUSSION THREAD

**Discussion by reece** [HGVSnomenclature/hgvs-nomenclature#111] [HVNC operations]
There have been several opinions about *how* we version. https://github.com/HGVSnomenclature/hgvs-nomenclature/discussions/12 received several comments that seemed to be in favor of "semantic versioning" and no one supporting date-based versioning, but I'm now aware that there are differing opinions. Let's resolve this once and for all.

Two options:

- Date based versioning -- What we've been doing.
- Semantic versioning -- Use versions like 1.2.3, where each position conveys meaning about the changes and compatibility with other versions.

**Comment by reece** [comment]
Here's my opinion. 

Date based versioning is insufficient to conveying which versions are compatible with others. That's a serious drawback for users. Sure, they can read the release notes, but that's a totally unsatisfactory outcome. Date based versioning is also incompatible with having multiple major versions at the same time. For example, we might want the conceptual equivalent of a 1.x and 2.x series, where 2.x is incompatible with 1.x, and we need to update both. Date-based versions can't handle that.

Conversely, semantic versioning handles this fairly neatly and conventionally. 


---

If we go with Semantic Versioning, then we need to pick a version number for the next release. We are migrating away from date-based versions toward semantic versioning-style versions. (see https://hgvs-nomenclature.org/en/latest/background/versioning/). The current version is 20.05.

Here are the options I considered and propose here for discussion:

- 20.06: This is the next obvious version, but isn't valid semver (leading 0).
- 20.6: Reasonable, but a weird jump from 20.05.
- 21.0.0: A major bump, but might imply to someone that it was released in 2021 and therefore not new.
- 24.0.0: Emphasizes new pages and resets the versioning scheme. Downside is that it uses a date when we say that we're moving away from dates.
- 24.1.0: I proposed this previously, but meant 24.0.0. I retract it. Sorry for the confusion.

My preference is 24.0.0

**Reply by murphyte** [reply]
> 21.0.0: A major bump, but might imply to someone that it was released in 2021 and therefore not new.

This is obviously a MAJOR version change. I'm not worried about the "21" being interpreted as 2021.

If we go with 24.0.0, then does that oblige us to switch to 25.0.0 on Jan 1 2025? That's a bit awkward to be forcing to do work (PR to increment the major version) on New Year's day every year forever more.

**Reply by reece** [reply]
Wait, you mean you think I was overthinking the interpretation of 21? Nah, couldn't be! :-)

**Reply by reece** [reply]
Also, just to be clear, it sounds like you're in favor of semantic versioning too. Right?

**Reply by ahwagner** [reply]
I am leaning away from semantic versioning for the HGVS Nomenclature; wrote this up as a separate comment thread, [here](https://github.com/HGVSnomenclature/hgvs-nomenclature/discussions/111#discussioncomment-7722986).

**Comment by ahwagner** [comment]
As someone who loves semantic versioning and uses it in all of his software projects (for the reasons @reece outlines [in his comment](https://github.com/HGVSnomenclature/hgvs-nomenclature/discussions/111#discussioncomment-7720103)), I find myself leaning away from that and towards calendar versioning for the HGVS Nomenclature.

I have two reasons for this. The first is that this specification is a combination of syntax and free-text rules for using and interpreting the syntax; for us to use semantic versioning, we would also need to create very clear rules about what what constitutes a _breaking change_, what is a _minor change_, and what is a _patch_, beyond whether or not the syntax changes. For example, if we revise the definition of a concept, is that a breaking change, even if we do not adjust the syntax? What if we revise the name of the concept? What if the definition changes are only clarifications / elaborations? What if these clarifications exclude some use cases that are syntactically valid but did not convey the intended message of that syntax with the clarifications? ... Not only do we need to answer these and myriad other questions for how to create version changes, but more importantly our community needs to intuitively understand how they are applied.

This brings me to my second point, the community. The group in this discussion thread (including me) is heavily biased towards variant model software developers / bioinformaticians. But when I hear from clinical groups that use HGVS talking about it they rarely discuss versions; they just reference whatever is on the website at the time. I have never read a paper where the HGVS version was supplied in the manuscript text, and would argue that if such papers exist they are the exception, not the rule. Every paper has a publication date, though! This makes it easy for our community to go back through calendar versions of HGVS to find what the current recommendations were at the time of the publication they are reading. In my view, the broadest part of the HGVS community has never even heard of Semantic Versioning; and so who are we serving if we go that route?

I think it would be more than appropriate for tools like Biocommons `hgvs` to use semantic versioning for parsing HGVS when interfaces to those tools need to change to accommodate revisions to the HGVS Nomenclature; very useful for software developers. But I am not sure that the HGVS Nomenclature would benefit in the way that we might hope it would if we made this switch.

Just my 2¢.

**Reply by reece** [reply]
For what it's worth, semantic versioning was favorably received at AMP. Two people mentioned it in the floor questions and comments, and two more privately afterward. One of the folks said that a dated version made it hard to know if it was the latest version.

No one spoke in support of dated versions. (But, I had just said that we were moving to semantic versioning, so why would they?)

**Reply by ahwagner** [reply]
I'm not sure how a semantic version helps in regard to knowing if they are looking at most recent, vs. a dated version. Don't both of these have a well-defined ordering? Regardless, RTD versioning tooling is very helpful for either, including a dynamic "latest" branch to address these concerns. It might be worthwhile formally polling relevant communities (e.g. ClinGen, AMP) with the question on versioning scheme if we're not already set on semantic versioning.

**Reply by reece** [reply]
Well, it's worth noting that that's the opposite of what you implied last year when this was being discussed.
https://github.com/HGVSnomenclature/hgvs-nomenclature/discussions/12#discussioncomment-4701852

**Reply by ahwagner** [reply]
I very much like semantic versioning, though to clarify, the comments from the thread last year are not for semantic versioning per se. They are for versioning all changes and for RTD tooling; I don’t think either of these is incompatible with calendar versioning. But I can see how that posts reads like an implicit endorsement for SemVer in HGVS Nomenclature, since the VRS example also uses SemVer. My other example (the fusions nomenclature) uses the tools I am proposing in that thread, but not SemVer. I apologize for any confusion this has caused!

I should also note that I do not have a strong preference for SemVer or Calendar, and I agree with the advantages of SemVer as @reece outlines them [below](https://github.com/HGVSnomenclature/hgvs-nomenclature/discussions/111#discussioncomment-7731881). Just laying out the advantages of the calendar system as I see them.

**Comment by reece** [comment]
> I'm not sure how a semantic version helps in regard to knowing if they are looking at most recent, vs. a dated version.

Well, I recommend that we focus on compatibility, not currency. Dated versions simply have no way to convey -- no semantics -- regarding compatibility. 

I understand your point re: what constitutes compatibility, but honestly this is always a judgement call, even in production systems. It's always theoretically possible that some minute patch-level-seeming change causes something to totally break in an unanticipated way. However, that should not -- indeed, for the sake of commercial end-users especially, must not -- keep us from aspiring to convey to people which changes are expected to be safe, which ones add new features, and which ones are likely to break compatibility. 


> Don't both of these have a well-defined ordering?

Not exactly. Semantic versioning is branchable and therefore a partial ordering across branches; dated versions are unbranchable and therefore a total ordering. 

In lay terms (for others on the thread), this means that with semantic versioning one can represent the idea that we have two major versions, 1.y.z and 2.y.z that are independent and can be evolved separately. This is essential for transition periods between changes. This idea is strictly impossible with time-based versions.

**Comment by reece** [comment]
Decided at HVNC meeting on 12/4: Semantic versioning, starting with 21.0.0.
